### PR TITLE
Fix test list filtering for CloudTest

### PR DESCRIPTION
**Pull request by karajas**
/cc @ChadNedzlek @MattGal

### CODE REVIEW

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+using System.Text;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    public sealed class FilterForUpload : Task
+    {
+        //This task is used to select the appropriate test zips from the different output directories produced by the build,
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: These should be proper XML doc comments. Just type "///" and VS will create the template for you

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            
+            foreach (var filename in filesToSearch)
+            {
+                string selectedTest = string.Empty;
+                using (StreamReader sr = new StreamReader(File.OpenRead(filename)))
+                {
+                    string line;
+                    while ((line = sr.ReadLine()) != null)
+                    {
+                        //Each line is of the format "$(BinDir)/tests/AnyOS.AnyCPU.Debug/Tests/netcoreapp1.0"
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Is this _literally_ "$(BinDir)" or is the expanded value?  If's the later, we'll have big problems, because my "BinDir" directory might contain all sorts of text you'll match on.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                        //Each test zip is of the format "$(BinDir)/tests/AnyOS.AnyCPU.Debug/archive/tests/netcoreapp1.0/Tests.zip"
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Is the test name not in the path anywhere? How do we upload multiple tests. If the test name _is_ in there somewhere, you can't use string contains, since the test name might itself contain things like "win"

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                }
+                if (!string.IsNullOrEmpty(selectedTest))
+                {
+                    this._copyFilteredUploadItems.Add(new TaskItem(selectedTest));
+                }
+            }
+            return true;
+        }
+        
+        private int calculateMatchValue(string lines)
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Casing, and "lines" isn't a good name for this parameter

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                    string line;
+                    while ((line = sr.ReadLine()) != null)
+                    {
+                        //Each line is of the format "$(BinDir)/tests/AnyOS.AnyCPU.Debug/Tests/netcoreapp1.0"
+                        //Each test zip is of the format "$(BinDir)/tests/AnyOS.AnyCPU.Debug/archive/tests/netcoreapp1.0/Tests.zip"
+                        //This is an implicit contract maintained in corefx 
+                        if (calculateMatchValue(line) > calculateMatchValue(selectedTest))
+                        {
+                            int lastIndexOfSlash = line.LastIndexOf('/');
+                            string parentDirectory = line.Substring(0, lastIndexOfSlash);
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Use Path.DirectoryName() instead of searching for the slashes yourself, especially since you use Path.Combine later.

File: src/Microsoft.DotNet.Build.Tasks/GenerateTestOutputDirList.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Xml.Linq;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class GenerateTestOutputDirList : Task
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.Tasks/GenerateTestOutputDirList.cs]: Nothing seems to use this Task. is it necessary? Or should it not be part of this change

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                                TestArchivesRoot);
+                            //Another contract in corefx is that the test name should be the AssemblyName appended with .Tests.
+                            string testZipName = parentDirectory.Substring(lastIndexOfSlash + 1) + ".zip";
+                            selectedTest = Path.Combine(testArchiveDir, testTFMFolderInArchive, testZipName);
+                        }
+                    }
+                }
+                if (!string.IsNullOrEmpty(selectedTest))
+                {
+                    this._copyFilteredUploadItems.Add(new TaskItem(selectedTest));
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: I think "this" is generally not used in CoreFX/BuildTools unless it's necessary (which it isn't here)

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 2 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: This is the expanded BinDir. I used this just for the comment.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        [Required]
+        public string FilterToOSGroup { get; set; }
+        //This is the path under a given Platform/OSGroup/Configuration folder where the Test Archives are placed
+        [Required]
+        public string TestArchivesRoot { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems
+        {
+            get { return this._copyFilteredUploadItems.ToArray(); }
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: This field should be named the same as the property (or, better yet, just us an auto-prop and assign it at the end of Execute)

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 3 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Here the capital Tests is stand in for an actual test name - Tests = Microsoft.CSharp.Tests.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 2 above)
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: In which case: fireworks. $(BinDir) could literally be any string under the sun that's a valid path, as we don't control where people clone the repo from. We need to only look specifically for the "AnyOS.AnyCPU.Debug" part and check that (preferably splitting it into it's sensible parts)

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 3 above)
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: See comment 3 lines up. We can't substring this then.

File: src/Microsoft.DotNet.Build.Tasks/GenerateTestOutputDirList.cs (hunk 6 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateTestOutputDirList.cs]: This is used in tests.targets to generate the list for each file.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 7 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 8 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 5 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 4 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 2 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved. I remove the value of BinDir from the generated test list during their creation.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 3 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        /// </summary>
+        /// This task is used to select the appropriate test zips from the different output directories produced by the build,
+        /// which are created based on OSGroup and TestTFM.
+        [Required]
+        public string TestOutputDirListsLocation { get; set; }
+
+        /// <summary>
+        /// The test tfm to select.
+        /// </summary>
+        /// This is the TestTFM folder that we want to select test zips from. This needs to be on the path to the test zip.
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Why are there two kinds of comments here?

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                {
+                    FilteredUploadItemsList.Add(new TaskItem(selectedTest));
+                }
+            }
+
+            FilteredUploadItems = FilteredUploadItemsList.ToArray();
+            return true;
+        }
+        
+        private int CalculateMatchValue(string testPathToMatch)
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: This still isn't safe, the "Example.Test" might be "CheckUnixWindows.Tests", which is going to do really, really bad things all over the place.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        public string FilterToOSGroup { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        private readonly List<ITaskItem> FilteredUploadItemsList = new List<ITaskItem>();
+
+        public override bool Execute()
+        {
+
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Still some really wonky formatting here. An extra blank line and some strange spaces around List&lt;ITaskItem>

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                    temp.Add(item);
+                }
+            }
+
+            Dictionary<string, List<ITaskItem>> map = new Dictionary<string, List<ITaskItem>>();
+            foreach (var item in temp)
+            {
+                string testName = Path.GetFileName(item.ItemSpec);
+                List<ITaskItem> values = null;
+                if (map.ContainsKey(testName))
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: We've talked about this. Don't do ContainsKey followed by TryGetValue, they literally do the same work twice.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+                {
+                    values = null;
+                    map.TryGetValue(testName, out values);
+                }
+                else
+                {
+                    values = new List<ITaskItem>();
+                }
+                values.Add(item);
+                map[testName] = values;
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: You shouldn't have to set this value except in the else (since it's already in the dictionary)

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        /// This is the OSGroup folder that we want to select test zips from. This does not require an exact match as some tests 
+        /// may or may not have an exact OSGroup folder match.
+        /// </summary>
+        [Required]
+        public string FilterToOSGroup { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        private readonly List<ITaskItem> FilteredUploadItemsList = new List<ITaskItem>();
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: This should be a local variable. And the name should reflect that.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    public sealed class FilterForUpload : Task
+    {
+        /// <summary>
+        /// This task is used to select the appropriate test zips from the different output directories produced by the build,
+        /// which are created based on OSGroup and TestTFM.
+        /// </summary>
+        [Required]
+        public ITaskItem[] UnfilteredUploadItems { get; set; }
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Something in the comment should indicate that we require the TestTFM and OSPlatformConfig properties to be set.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        public string BinDir { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        public override bool Execute()
+        {
+            List<ITaskItem> filteredBasedOnTestTfm =
+                UnfilteredUploadItems.Where(
+                    item => item.ItemSpec.Replace(BinDir, string.Empty).Contains(FilterToTestTFM)).ToList();
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Usually "Replace" when trying to make files relative is a hack, because whatever you are attempting to make it relative to _might_ appear a second time in the path ("../foo/pizza/pie/../foo" and you're trying to remove "../foo", it will mutilate that path).

Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Noted. The path in the ItemSpec is constructed using the BinDir here though.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+            List<ITaskItem> localFilteredUploadItemsList = new List<ITaskItem>();
+
+            Dictionary<string, List<ITaskItem>> map = new Dictionary<string, List<ITaskItem>>();
+            foreach (var item in filteredBasedOnTestTfm)
+            {
+                string testName = Path.GetFileName(item.ItemSpec);
+                List<ITaskItem> values = null;
+                map.TryGetValue(testName, out values);
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: if (!map.TryGetValue(testName, out values))

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 16 above)
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: Use this instead of "thing.Replace"

``` csharp
string myRandomPath = ...;
string binDir = ...;

binDir = Path.GetFullPath(binDir);
string fullRandomPath = Path.GetFullPath(myRandomPath);
if (fullRandomPath.StartsWith(binDir)) {
  return fullRandomPath.Substring(binDir.Length + 1)
else
  return myRandomPath;
```

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+                }
+                localFilteredUploadItemsList.Add(selectedItem);
+            }
+
+            FilteredUploadItems = localFilteredUploadItemsList.ToArray();
+            return true;
+        }
+
+
+        private string makeRelativePathTo(string path1, string path2)
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: makeRelativePathTo -> MakeRelativePathTo
path1 -> targetPath
path2 -> basePath

**Comment by ChadNedzlek** [comment]
One last tiny change, then squash, and then, assuming we got a good job with the correct number of tests in it (about 160ish) it will look good to me.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    public sealed class FilterForUpload : Task
```
Comment (tarekgh) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: > public sealed class FilterForUpload : Task [](start = 3, length = 43)

could you please add a comment describing what this task for and how it work?

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets [hunk 20]
```diff
@@ -194,11 +209,9 @@
...
           Condition="'$(Performance)' != 'true' or '$(FuncTestsDisabled)' != 'true'">
     <!-- create item group of functional tests -->
-    <ItemGroup>
-      <FunctionalTest Include="$(TestArchivesRoot)**/*.zip" />
-      <FunctionalTest Include="$(AnyOSTestArchivesRoot)**/*.zip" />
-      <FunctionalTest Condition="'$(TargetsUnix)' == 'true'" Include="$(UnixTestArchivesRoot)**/*.zip" />
-    </ItemGroup>
+    <CreateItem Include="$(ForUploadList)">
+      <Output TaskParameter="Include" ItemName="FunctionalTest"/>
+    </CreateItem>
```
Comment (tarekgh) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: why we didn't do 

<ItemGroup>
  <FunctionalTest Include="$(ForUploadList)" />
</ItemGroup>

Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: It's using the same output of the previous one - ForUploadList?

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+            FilteredUploadItems = localFilteredUploadItemsList.ToArray();
+            return true;
+        }
+
+
+        private string makeRelativePathTo(string path1, string path2)
+        {
+            path2 = Path.GetFullPath(path2);
+            string fullPath1 = Path.GetFullPath(path1);
+            if (fullPath1.StartsWith(path2))
```
Comment (tarekgh) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: >ullPath1.StartsWith(path2)) [](start = 17, length = 27)

make thins ordinal ignore case to avoid running into weird cases

Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: I agree that's probably ok, since we can't detect the case-sensitivity of the underlying OS.  In general, I think our build tries hard not to have thing meaningfully differ by case, so it's probably safe.

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+            foreach (var item in map)
+            {
+                List<ITaskItem> values = item.Value;
+                ITaskItem selectedItem = null;
+                foreach (var value in values)
+                {
+                    if (value.GetMetadata("OSPlatformConfig").Contains(FilterToOSGroup))
+                    {
+                        selectedItem = value;
```
Comment (tarekgh) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: >selectedItem = value; [](start = 24, length = 21)

should we break the loop if we find selectedItem?

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets [hunk 23]
```diff
@@ -89,14 +90,28 @@
...
+      </UnfilteredTestArchives>
       <!-- Only include Unix folders if supported by the current target OS -->
-      <ForUpload Condition="'$(TargetsUnix)' == 'true'" Include="$(UnixTestArchivesRoot)**/*.zip" />
+      <UnfilteredTestArchives Condition="'$(TargetsUnix)' == 'true'" Include="$(UnixTestArchivesRoot)**/*.zip" >
+        <OSPlatformConfig>$(UnixPlatformConfig)</OSPlatformConfig>
+      </UnfilteredTestArchives>
     </ItemGroup>
+    <FilterForUpload UnfilteredUploadItems="@(UnfilteredTestArchives)"
+             FilterToTestTFM="$(FilterToTestTFM)"
+             FilterToOSGroup="$(FilterToOSGroup)"
```
Comment (tarekgh) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: can we add like <Error ..> with the condition assert FilterToTestTFM/OSGroup are not empty. I assume this should be the case. right?

File: src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs (hunk 22 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs]: resolved

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets (hunk 20 above)
Comment (tarekgh) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: I am ok to keep what you have but you can empty the previous list if you want. 

<FunctionalTest Remove="@(FunctionalTest )"/>


---
In reply to: [70545684](https://github.com/dotnet/buildtools/pull/870#discussion_r70545684) [](ancestors = 70545684)

**Comment by tarekgh** [comment]
:shipit:

**Comment by ChadNedzlek** [comment]
:shipit:
